Import/export command refactoring

emilmelnikov opened the pull request.

btbest commented:
Small unrelated thing I would squeeze in while you're pushing out an advertised update: changing the path hints in ui\ilastikOptions.java from `ilastik-1.3.3` to `ilastik-1.4.0` 🙃

btbest commented on src/main/java/org/ilastik/ilastik4ij/io/ListDatasetsCommand.java:
```diff
@@ -0,0 +1,45 @@
+
+import static org.scijava.ItemIO.OUTPUT;
+
+@Plugin(type = Command.class, headless = true, menuPath = "Plugins>ilastik>List HDF5 Datasets")
```
```suggestion
@Plugin(type = Command.class, headless = true, menuPath = "Plugins>ilastik>List datasets in an HDF5 file")
```

emilmelnikov commented:
> Small unrelated thing I would squeeze in while you're pushing out an update: changing the path hints in ui\ilastikOptions.java from `ilastik-1.3.3` to `ilastik-1.4.0` 🙃

What do you think about trying to autodetect ilastik path and, if successful, use it as a default value (if there is no existing value of course)? For macOS, it's almost always `/Applications/Fiji.app`, for Windows it's probably something like `C:\Program Files\Fiji`, for Linux it's not obvious, but we can still try to autodetect this.

btbest commented:
> for Windows it's probably something like C:\Program Files\Fiji

Fiji explicitly advises against unzipping it to C:\Program Files because it can't self-update if it's there. So if people take this advice to herat, Fiji should have no default location on Windows besides possibly `%userprofile%\Downloads\Fiji.app` 🙈

emilmelnikov commented:
> > for Windows it's probably something like C:\Program Files\Fiji
> 
> Fiji explicitly advises against unzipping it to C:\Program Files because it can't self-update if it's there. So if people take this advice to heart, Fiji should have no default location on Windows besides possibly `%userprofile%\Downloads\Fiji.app` 🙈 But we are using `C:\Program Files` as an example path right now anyway... how are you thinking to autodetect it, given it's a portable app?

Oh, sorry, the examples where totally wrong, I meant `/Applications/ilastik.app`, `C:\Program Files\ilastik` etc.

btbest commented:
Oh right,and I clearly got just as confused... Yeah, in fact I had just the same thought whether instead of determining only the path hint, we could also just use that hint as the default value for the field.

k-dominik commented:
Let's not add any autodetection of paths to this PR :). I think in the past we thought about writing some file (from within ilastik) to the home folder that would point to the ilastik installation...

btbest commented on src/main/java/org/ilastik/ilastik4ij/io/ExportCommand.java:
```diff
@@ -0,0 +1,80 @@
+        ImgPlus<T> img = (ImgPlus<T>) input.getImgPlus();
+        List<AxisType> axes = toImagejAxes(reversed(AXIS_ORDER));
+
+        long totalBytes = img.size() * img.firstElement().getBitsPerPixel() / 8;
+        if (totalBytes >> 20 > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("Image is too large to export");
+        }
+        int totalMegabytes = (int) (totalBytes >> 20);
```
Could pull this bit out into a method `getImageSize` for readability

btbest reviewed (approved):
Looks good to merge for me, nothing substantial in my suggestions 🚀

emilmelnikov commented on src/main/java/org/ilastik/ilastik4ij/io/ListDatasetsCommand.java:
```diff
@@ -0,0 +1,45 @@
+
+import static org.scijava.ItemIO.OUTPUT;
+
+@Plugin(type = Command.class, headless = true, menuPath = "Plugins>ilastik>List HDF5 Datasets")
```
Note: we won't be able to change this after the release (title is used in macros), so we should be sure about the naming!

btbest replied:
Oof, I didn't realise that. In that case I think I actually lean towards "List HDF5 Datasets" as you had put it.

k-dominik reviewed (commented):
Hi @emilmelnikov,

i didn't look too much into the code, but I really like that we got rid of the custom dialogs and are using the buildin plugin machinery for that. The interactivity you added to the import dialog is a really nice feature.

I tried the macros. In gui mode they work very well (as before). About the headless claim I'm not so sure. While there is definitely something running (including ilastik) it is impossible to do something with this data later in the macro (this was also the case before). _I don't think this can be fixed easily_.

Minor issue:
* if the import command fails, the macro continues running, an error pops up though. An error in `open` for example will cause the macro to abort (which would be desirable I think).

btbest commented on README.md:
```diff
@@ -255,33 +255,15 @@ RandomAccessibleInterval`.
 
-Make sure that the released plugin version is not installed in Fiji, and that
-`local-install-to-fiji` profile is not disabled.
+Run `scripts/fiji-deploy.py` to build project and copy the newly built plugin to Fiji.
```
In my case it didn't work because I installed maven over conda :grin: So the subprocess doesn't have a valid `mvn` command. But the script contains what I need to do manually, so it still made installing the plugin from code easy and quick for me.

k-dominik replied:
I also installed maven via conda - for me it works :)

emilmelnikov commented:
About options dialog: I've updated version numbers, and also moved example path into a separate line:
<img width="1148" alt="image" src="https://github.com/ilastik/ilastik4ij/assets/1649961/de44c315-e4a4-449b-9fd3-9d7f18cbcd40">

Another point: on macOS, we currently switch to directory style file picker, but for some reason it doesn't allow a user to pick `.app` files. Should we ditch directory style picker and ask users to select `Contents/MacOS/ilastik` explicitly? This requires more clicks, but least it's actually pickable, and with the example path it should be pretty obvious what to do.

emilmelnikov commented on examples/import.ijm:
```diff
@@ -0,0 +1,5 @@
+input = "src/test/resources/2d_cells_apoptotic_1channel.h5";
```
In this case, I think it's better to tell users to provide their own parameters explicitly, maybe with some dialog fallback, like this:

```
#@ File (label = "Input file", style = "file") input
// input = "/path/to/some/directory/src/test/resources/2d_cells_apoptotic.tif";

#@ File (label = "Output file", style = "save") exportPath
// exportPath = "/path/to/some/directory/output.h5";

datasetName = "/data";
compressionLevel = 0;

run("Export HDF5", "input=[" + input + "] datasetname=[" + datasetName + "] compressionlevel=[" + compressionLevel + "] exportpath=[" + exportPath + "]");
```

Do you think it's self-explanatory enough?

k-dominik replied:
yeah that should suffice. Note - with those parameter thingies, headless mode does not work at all :bummer:, so at least for the export command, we should not do it :)

emilmelnikov replied:
Yes, this was the intention: use interactive dialogs by default, and show how to set the same parameters explicitly in commented-out lines. Do you think it needs to be more explicit? How much familiarity with the macro language should we expect from our users?..

k-dominik replied:
You're right of course. Maybe let's be extra nice and _only for the export macro (bc others don't really run headles)_ add a comment that the parameters can be removed, specified in code directly for headless operation.

emilmelnikov replied:
Added the following description in `export.ijm`:

```
// Lines that start with "//" are comments.

// Lines that start with "#@" declare input parameters that interactively ask the user for values.
// If you want to run a macro in batch or headless mode, remove those lines and set input parameters explicitly.
```

emilmelnikov commented on examples/object_classification.ijm:
```diff
@@ -0,0 +1,7 @@
+secondInput = "src/test/resources/2d_cells_apoptotic_1channel-data_Probabilities.tif";
+type = "Probabilities";
+
+open(input);
```
I think you have to open at least one image, otherwise Fiji complains that there are no opened datasets. Not sure why it works without opening the second image...

Also, if I try to open the second input, Fiji shows Bio-formats import dialog. I tried to re-save the probabilities TIFF after importing, but it saves it in `XYC` axis order (channels last), and this trips up our workflow.

emilmelnikov commented on README.md:
````diff
@@ -255,33 +255,15 @@ RandomAccessibleInterval`.
-    </profiles>
-</settings>
-```
+It's possible to run some example macros from command line with
+`/path/to/fiji/executable --headless -macro examples/your-example.ijm`.
+Note that headless mode behaves differently than the GUI.
````
Issues with paths should be fixed by requiring absolute paths in all cases (either via dialogs or by specifying them explicitly).

Output from `object_classification.ijm` looks weird because, for some reason, it outputs _Simple Segmentation_ instead of _Probabilites_ (you can see it by applying _Process_ -> _Enhance Contrast..._).

In any case, I agree that we shouldn't mention headless mode in README. I added this just to note that these 2 modes could work differently in general, but it is indeed confusing.

emilmelnikov replied:
For `convert_tiff_to_ilastik_h5.ijm`, I didn't get any popup dialogs except for that Bioformats import dialog (I tried to run it on `src/test/resources`).

emilmelnikov replied:
Incorrect output in `object_classification.ijm` was caused by a bug: we didn't pass `--export_source` in `ObjectClassificationWorkflow`. Fixed this.

k-dominik reviewed (commented):
I've been going over the changes mostly from a user perspective and it seems pretty much compatible with the last version, that's awesome! And so many ux improvements :)

One thing I noticed is that on import, the name of the image is different. In the old version of the plugin, it would be External/Internal path, now, only the Internal path shows up as image name. It would be really nice to have this behavior not change (some macros might rely on selecting the image that way). Also I think it's probably helpful to see the full path, as two different images might easily be stored in the same internal path in different files. Note, this seems to only occur when using the "import" functionality. When running a workflow, the full path will be displayed on the result image.

emilmelnikov commented:
> In the old version of the plugin, it would be External/Internal path, now, only the Internal path shows up as image name.

It was a silly bug, sorry about that. This has been fixed.

emilmelnikov commented:
What do you think about shortening labels in the options dialod?

<img width="740" alt="image" src="https://github.com/ilastik/ilastik4ij/assets/1649961/10fd0ea4-fc3f-4f85-b85c-a1fa18727978">

Longer explanations could be moved into _description_ fields, they are showed as tooltops in the UI.

k-dominik commented:
I like the shortened version more, a bit of a pity that the fields for memory and threads are so wide, but I get it, the grid matches the one for filename... I'm fine if they show up on tooltip. I'd add that Memory is in MB (would actually be nicer to GB there, I mean who will bargain over fractions of that?), and it would be cool to add something about -1.. Maybe Threads (-1 means no restriction), or even shorter.

emilmelnikov commented:
> I like the shortened version more, a bit of a pity that the fields for memory and threads are so wide, but I get it, the grid matches the one for filename... I'm fine if they show up on tooltip. I'd add that Memory is in MB (would actually be nicer to GB there, I mean who will bargain over fractions of that?), and it would be cool to add something about -1.. Maybe Threads (-1 means no restriction), or even shorter.

<img width="764" alt="image" src="https://github.com/ilastik/ilastik4ij/assets/1649961/45cc4fdc-4386-447d-afad-2915043ba501">
